File: src/api/spider.ts
import { api } from '@/lib/api';

export interface SpiderPlatform {
  id: string;
  name: string;
  display_name: string;
  status: 'active' | 'paused' | 'error';
  reliability: number;
  error_log?: string;
  config: any;
  last_sync_at?: string;
  created_at: string;
}

export interface SpiderSyncLog {
  id: string;
  platform_id: string;
  store_id: string;
  status: 'running' | 'success' | 'failed';
  records_synced: number;
  error_message?: string;
  duration_ms: number;
  started_at: string;
  finished_at?: string;
}

export interface SpiderTask {
  id: string;
  platform_id: string;
  store_id: string;
  task_type: 'full_sync' | 'incremental' | 'reply';
  status: 'pending' | 'running' | 'success' | 'failed';
  priority: number;
  result: any;
  error_message?: string;
  scheduled_at?: string;
  started_at?: string;
  finished_at?: string;
}

export const spiderApi = {
  getPlatforms: async (): Promise<SpiderPlatform[]> => {
    const response = await api.get<any>('/v1/spider/platforms');
    return response.data || response;
  },

  createPlatform: async (data: Partial<SpiderPlatform>): Promise<SpiderPlatform> => {
    const response = await api.post<any, any>('/v1/spider/platforms', data);
    return response.data || response;
  },

  updatePlatform: async (id: string, data: Partial<SpiderPlatform>): Promise<SpiderPlatform> => {
    const response = await api.put<any, any>(`/v1/spider/platforms/${id}`, data);
    return response.data || response;
  },

  deletePlatform: async (id: string): Promise<void> => {
    const response = await api.delete(`/v1/spider/platforms/${id}`);
    return response.data;
  },

  syncPlatform: async (id: string): Promise<void> => {
    const response = await api.post(`/v1/spider/platforms/${id}/sync`);
    return response.data;
  },

  syncAllPlatforms: async (): Promise<void> => {
    const response = await api.post('/v1/spider/sync-all');
    return response.data;
  },

  getSyncLogs: async (platformId?: string, page: number = 1, pageSize: number = 20): Promise<any> => {
    let url = `/v1/spider/logs?page=${page}&page_size=${pageSize}`;
    if (platformId) url += `&platform_id=${platformId}`;
    const response = await api.get<any>(url);
    return response.data || response;
  },

  testPlatformConnection: async (id: string): Promise<void> => {
    const response = await api.post(`/v1/spider/platforms/${id}/test`);
    return response.data;
  },

  getTasks: async (status?: string): Promise<SpiderTask[]> => {
    const url = status ? `/v1/spider/tasks?status=${status}` : '/v1/spider/tasks';
    const response = await api.get<any>(url);
    return response.data || response;
  },

  createTask: async (data: Partial<SpiderTask>): Promise<SpiderTask> => {
    const response = await api.post<any, any>('/v1/spider/tasks', data);
    return response.data || response;
  },

  cancelTask: async (id: string): Promise<void> => {
    const response = await api.post(`/v1/spider/tasks/${id}/cancel`);
    return response.data;
  },
};

File: backend/app/api/v1/spider.py
"""
爬虫路由模块
处理爬虫平台管理、同步日志、任务管理等接口
"""

from typing import Optional
from uuid import UUID

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.deps import require_valid_subscription, get_db
from app.core.response import paginated, success
from app.models.user import User
from app.schemas.spider import (
    SpiderPlatformCreateRequest,
    SpiderPlatformListResponse,
    SpiderPlatformResponse,
    SpiderPlatformUpdateRequest,
    SpiderSyncLogListResponse,
    SpiderSyncLogResponse,
    SpiderTaskCreateRequest,
    SpiderTaskListResponse,
    SpiderTaskResponse,
    SpiderTestConnectionResponse,
)
from app.services import spider_service

router = APIRouter(prefix="/spider", tags=["爬虫管理"])


@router.get("/platforms", summary="平台列表")
async def get_platforms(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_valid_subscription),
) -> dict:
    """
    获取所有爬虫平台列表
    """
    platforms = await spider_service.get_platforms(db)
    return success(
        data=[
            SpiderPlatformResponse.model_validate(p).model_dump(mode="json")
            for p in platforms
        ]
    )


@router.post("/platforms", summary="新增平台")
async def create_platform(
    request: SpiderPlatformCreateRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_valid_subscription),
) -> dict:
    """
    创建新的爬虫平台
    """
    platform = await spider_service.create_platform(db, request)
    return success(
        data=SpiderPlatformResponse.model_validate(platform).model_dump(mode="json"),
        message="平台创建成功",
    )


@router.put("/platforms/{platform_id}", summary="更新平台")
async def update_platform(
    platform_id: UUID,
    request: SpiderPlatformUpdateRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_valid_subscription),
) -> dict:
    """
    更新爬虫平台信息
    """
    platform = await spider_service.update_platform(db, platform_id, request)
    return success(
        data=SpiderPlatformResponse.model_validate(platform).model_dump(mode="json"),
        message="平台更新成功",
    )


@router.delete("/platforms/{platform_id}", summary="删除平台")
async def delete_platform(
    platform_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_valid_subscription),
) -> dict:
    """
    删除爬虫平台
    """
    await spider_service.delete_platform(db, platform_id)
    return success(message="平台删除成功")


@router.post("/platforms/{platform_id}/sync", summary="同步单个平台")
async def sync_platform(
    platform_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_valid_subscription),
) -> dict:
    """
    手动触发单个平台的数据同步
    """
    sync_log = await spider_service.sync_platform(db, platform_id)
    return success(
        data=SpiderSyncLogResponse.model_validate(sync_log).model_dump(mode="json"),
        message="同步任务已启动",
    )


@router.post("/sync-all", summary="同步所有平台")
async def sync_all_platforms(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_valid_subscription),
) -> dict:
    """
    手动触发所有活跃平台的数据同步
    """
    sync_logs = await spider_service.sync_all_platforms(db)
    return success(
        data=[
            SpiderSyncLogResponse.model_validate(log).model_dump(mode="json")
            for log in sync_logs
        ],
        message=f"已启动 {len(sync_logs)} 个平台的同步任务",
    )


@router.get("/logs", summary="同步日志")
async def get_sync_logs(
    platform_id: Optional[UUID] = Query(None, description="平台ID筛选"),
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页数量"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_valid_subscription),
) -> dict:
    """
    获取爬虫同步日志列表
    """
    logs, total = await spider_service.get_sync_logs(db, platform_id, page, page_size)
    return paginated(
        items=[
            SpiderSyncLogResponse.model_validate(log).model_dump(mode="json")
            for log in logs
        ],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.post("/platforms/{platform_id}/test", summary="测试连接")
async def test_platform_connection(
    platform_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_valid_subscription),
) -> dict:
    """
    测试爬虫平台连接
    """
    result = await spider_service.test_platform_connection(db, platform_id)
    return success(data=result)


@router.get("/tasks", summary="任务列表")
async def get_tasks(
    status: Optional[str] = Query(None, description="状态筛选: pending/running/success/failed"),
    page: int = Query(1, ge=1, description="页码"),
    page_size: int = Query(20, ge=1, le=100, description="每页数量"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_valid_subscription),
) -> dict:
    """
    获取爬虫任务列表
    """
    tasks, total = await spider_service.get_tasks(db, status, page, page_size)
    return paginated(
        items=[
            SpiderTaskResponse.model_validate(task).model_dump(mode="json")
            for task in tasks
        ],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.post("/tasks", summary="创建任务")
async def create_task(
    request: SpiderTaskCreateRequest,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_valid_subscription),
) -> dict:
    """
    创建新的爬虫任务
    """
    task = await spider_service.create_task(db, request)
    return success(
        data=SpiderTaskResponse.model_validate(task).model_dump(mode="json"),
        message="任务创建成功",
    )


@router.post("/tasks/{task_id}/cancel", summary="取消任务")
async def cancel_task(
    task_id: int,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(require_valid_subscription),
) -> dict:
    """
    取消待执行或运行中的爬虫任务
    """
    task = await spider_service.cancel_task(db, task_id)
    return success(
        data=SpiderTaskResponse.model_validate(task).model_dump(mode="json"),
        message="任务已取消",
    )
